src/script/interpreter.cpp

```diff
@@ -1264,6 +1290,179 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    //
+                    //   ( <trigger-spk-hash> | <target-hash> )
+                    //   <spend-delay>
+                    //   <recovery-params>
+                    //   <recovery-vout-idx>
+                    //
+                    if (stack.size() < 4) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    int recovery_vout_idx;
+                    try {
+                        recovery_vout_idx = CScriptNum(stacktop(-4), fRequireMinimal).getint();
+                    } catch (...) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_BAD_VOUT_IDX);
+                    }
+                    const bool is_recovery = recovery_vout_idx >= 0;
+
+                    // Note that this is making a copy, not taking a reference, so that
+                    // we avoid UB when using this state after stack pops.
+                    const valtype recovery_params = stacktop(-3);
+                    if (recovery_params.size() < uint256::WIDTH) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_INVALID_RECOVERY_PARAMS);
+                    }
+
+                    // `spend-delay` is a CScriptNum, up to 4 bytes, that is interpreted
+                    // in the same way as the first 23 bits of nSequence are per
+                    // BIP 68 (relative time-locks). This enables users of vaults to
+                    // express spend delays in either wall time or block count, and reuse
+                    // the same machinery as OP_CHECKSEQUENCEVERIFY.
+                    const CScriptNum spend_delay(stacktop(-2), fRequireMinimal);
+                    if (spend_delay < 0) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_INVALID_DELAY);
+                    }
+
+                    const valtype& hash_bytes_from_stack = stacktop(-1);
+                    if (hash_bytes_from_stack.size() != 32) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+                    // Hash value that will be interpreted based on OP_VAULT vs. OP_UNVAULT.
+                    const uint256 triggerspk_or_target_hash{hash_bytes_from_stack};
+
+                    popstack(stack);
+                    popstack(stack);
+                    popstack(stack);
+                    popstack(stack);
+
+                    // Case 1 for OP_VAULT/OP_UNVAULT: sweep to recovery
+                    if (is_recovery) {
+                        CScript recovery_spk;
+
+                        if (const auto& err = checker.CheckVaultSpendToRecoveryOutputs(
+                                recovery_vout_idx, execdata, recovery_params, recovery_spk, flags)) {
+                            return set_error(serror, *err);
+                        }
+                        // If an optional recovery authorization witness program has been given,
```

Comment:
From my understanding, the optional recovery authorisation witness program is included in the OP_VAULT `recovery_params`. Therefore in case of future script extensions deployed a posteriori of the OP_VAULT commitment, one has to re-spend your old OP_VAULT output to a new one to benefit from script extensions. I think an alternative could be to have a `recovery_pubkey` blessing the new scripts, in the way grafroot envisioned it: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-February/015700.html

Reply:
I think graftroot-like key delegation is probably out of scope for this proposal; such a technique would be useful for many other parts of bitcoin, and if it is implemented, it could be used here too since the recovery params are compatible with future witness verisons. In the meantime if you want to change the recovery rules for a given vault, I think withdrawaling it into a new vault is a reasonable solution.

Reply:
Sure the recovery params are compatible with future witness versions, however a specific witness version is included as part of the recovery params hash (e.g version 2 or version 3 witness program) iiuc. Therefore if you would like to upgrade your recovery authorization witness program towards a new witness program you have to unvault the funds towards a new vault instance.

Instead, if you have a grafroot-like key delegation you just have to re-sign the surrogate scripts, I think. Of course, it doesn't invalidate the old surrogates that might present now a safety downgrade from your vault policy. However, this solution could be batched to all your vaults under custody, as they can all share the grafroot-like key ? So sounds a trade-off between operational complexity and introducing a synchronicity point if you have to withdraw all the funds to a new vault if you would like to change the recovery rules.